rcl/test/rcl/test_parameters.cpp

```diff
@@ -0,0 +1,383 @@
+
+#include "rcl/rcl.h"
+
+#include "../memory_tools/memory_tools.hpp"
+#include "../scope_exit.hpp"
+
+#ifdef RMW_IMPLEMENTATION
+# define CLASSNAME_(NAME, SUFFIX) NAME ## __ ## SUFFIX
+# define CLASSNAME(NAME, SUFFIX) CLASSNAME_(NAME, SUFFIX)
+#else
+# define CLASSNAME(NAME, SUFFIX) NAME
+#endif
+
+
+class CLASSNAME(TestParametersFixture, RMW_IMPLEMENTATION) : public ::testing::Test
+{
+public:
+  rcl_node_t * node_ptr;
+  rcl_wait_set_t * wait_set;
+  rcl_parameter_service_t * parameter_service;
+  rcl_parameter_client_t * parameter_client;
+  const size_t num_params = 5;
+
+  void SetUp()
+  {
+    stop_memory_checking();
+    rcl_ret_t ret;
+    ret = rcl_init(0, nullptr, rcl_get_default_allocator());
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+    this->node_ptr = new rcl_node_t;
+    *this->node_ptr = rcl_get_zero_initialized_node();
+    const char * name = "parameter_node";
+    rcl_node_options_t node_options = rcl_node_get_default_options();
+    ret = rcl_node_init(this->node_ptr, name, &node_options);
+
+    ret = rcl_wait_set_init(this->wait_set, 0, 0, 0, 0, 0, rcl_get_default_allocator());
+
+    this->parameter_service = new rcl_parameter_service_t;
+    *this->parameter_service = rcl_get_zero_initialized_parameter_service();
+    rcl_parameter_service_options_t ps_options = rcl_parameter_service_get_default_options();
+    ret = rcl_parameter_service_init(this->parameter_service, this->node_ptr, &ps_options);
+
+    this->parameter_client = new rcl_parameter_client_t;
+    *this->parameter_client = rcl_get_zero_initialized_parameter_client();
+    rcl_parameter_client_options_t cs_options = rcl_parameter_client_get_default_options();
+    ret = rcl_parameter_client_init(this->parameter_client, this->node_ptr, &cs_options);
+
+    // These implicitly add the parameter clients, services and subscriptions to the waitset
+    ret = rcl_wait_set_add_parameter_client(wait_set, this->parameter_client);
+    ret = rcl_wait_set_add_parameter_service(wait_set, this->parameter_service);
+
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+    set_on_unexpected_malloc_callback([]() {ASSERT_FALSE(true) << "UNEXPECTED MALLOC";});
+    set_on_unexpected_realloc_callback([]() {ASSERT_FALSE(true) << "UNEXPECTED REALLOC";});
+    set_on_unexpected_free_callback([]() {ASSERT_FALSE(true) << "UNEXPECTED FREE";});
+    start_memory_checking();
+  }
+
+  void TearDown()
+  {
+    assert_no_malloc_end();
+    assert_no_realloc_end();
+    assert_no_free_end();
+    stop_memory_checking();
+    set_on_unexpected_malloc_callback(nullptr);
+    set_on_unexpected_realloc_callback(nullptr);
+    set_on_unexpected_free_callback(nullptr);
+    rcl_ret_t ret = rcl_node_fini(this->node_ptr);
+    delete this->node_ptr;
+
+
+    ret = rcl_wait_set_fini(this->wait_set);
+    delete this->wait_set;
+
+    ret = rcl_parameter_service_fini(this->parameter_service);
+    delete this->parameter_service;
```

Comment:
You should probably delete these in reverse order, i.e. you should probably fini and delete the parameter service before the node.